src/compas/datastructures/mesh/mesh.py

```diff
@@ -486,7 +491,36 @@ def to_vertices_and_faces(self):
         """
         key_index = self.key_index()
         vertices = [self.vertex_coordinates(key) for key in self.vertices()]
-        faces = [[key_index[key] for key in self.face_vertices(fkey)] for fkey in self.faces()]
+
+        if not triangulated:
+            faces = [[key_index[key] for key in self.face_vertices(fkey)] for fkey in self.faces()]
+            return vertices, faces
+
+        faces = []
+
+        for fkey in self.faces():
+            face_vertices = self.face_vertices(fkey)
+
+            if len(face_vertices) == 3:
+                a, b, c = face_vertices
+                faces.append([key_index[a], key_index[b], key_index[c]])
+            elif len(face_vertices) == 4:
+                a, b, c, d = face_vertices
+                faces.append([key_index[a], key_index[b], key_index[c]])
+                faces.append([key_index[a], key_index[c], key_index[d]])
+            else:
+                centroid = centroid_polygon([self.vertex_coordinates(key) for key in face_vertices])
+                vertices.append(centroid)
+                vcount = len(vertices)
+
+                for i in range(len(face_vertices) - 1):
+                    a, b, c = face_vertices[i], face_vertices[i + 1], vcount - 1
+                    faces.append([key_index[a], key_index[b], c])
+
+                # last face between first, last and centroid
+                a, b, c = face_vertices[0], face_vertices[-1], vcount - 1
+                faces.append([key_index[b], key_index[a], c])
```

Comment:
this is what pairwise is for
```python
for a, b in pairwise(face_vertices + face_vertices[:1]):
    faces.append([key_index[a], key_index[b], vcount - 1])
```

also, if you do `len(vertices)` before appending the centroid, you can avoid the repeated `vcount - 1`

Reply:
fixed at https://github.com/compas-dev/compas/pull/1013/commits/dd3f87fc64a4038874a51cfd8a95ee738ed318bd